Make step() faster by passing in a tensor vs scalar 1

janeyx99 opened the pull request:
This is the culminated result of https://github.com/pytorch/pytorch/pull/110954#issuecomment-1758520411.

We are making the code slightly more complicated to gain some perf in minimizing calls to `.copy_()` and `.to()`.

### Code
```
import torch
with torch.cuda.device(0):
    steps = [torch.zeros((), device="cpu", dtype=torch.float32) for i in range(1000)]

    with torch.profiler.profile(
        activities=[
            torch.profiler.ProfilerActivity.CPU,
            torch.profiler.ProfilerActivity.CUDA,
        ]
    ) as p:
        # New code:
        # step_device = steps[0].device
        # one = torch.tensor(1.0, device=step_device) if str(step_device) == "cpu" else 1
        # torch._foreach_add_(steps, one, 1.0)

        # Old code:
        torch._foreach_add_(steps, 1)

    print(p.key_averages().table(sort_by="cpu_time_total"))
```

### Profiles
**with old code**
```
-------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
                     Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg    # of Calls  
-------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
      aten::_foreach_add_        35.31%      52.089ms        99.99%     147.495ms     147.495ms             1  
               aten::add_        25.05%      36.949ms        64.68%      95.406ms      95.406us          1000  
                 aten::to         3.97%       5.852ms        39.63%      58.457ms      58.457us          1000  
           aten::_to_copy        10.11%      14.917ms        35.66%      52.605ms      52.605us          1000  
              aten::copy_        21.65%      31.939ms        21.65%      31.939ms      31.939us          1000  
      aten::empty_strided         3.90%       5.749ms         3.90%       5.749ms       5.749us          1000  
    cudaDeviceSynchronize         0.01%      18.000us         0.01%      18.000us      18.000us             1  
-------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
Self CPU time total: 147.513ms
```

**with new code**
```
-------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
                     Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg    # of Calls  
-------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
      aten::_foreach_add_        55.06%      49.963ms        99.86%      90.625ms      90.625ms             1  
               aten::add_        44.81%      40.662ms        44.81%      40.662ms      40.662us          1000  
            aten::detach_         0.01%       8.000us         0.05%      45.000us      45.000us             1  
                  detach_         0.04%      37.000us         0.04%      37.000us      37.000us             1  
              aten::empty         0.03%      30.000us         0.03%      30.000us      30.000us             1  
                 aten::to         0.03%      23.000us         0.03%      23.000us      23.000us             1  
    cudaDeviceSynchronize         0.02%      22.000us         0.02%      22.000us      22.000us             1  
         aten::lift_fresh         0.01%       6.000us         0.01%       6.000us       6.000us             1  
-------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
Self CPU time total: 90.751ms
```

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #111600
* __->__ #111084
* #111079



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

janeyx99 committed:
Make step() faster by passing in a tensor vs scalar 1

[ghstack-poisoned]

janeyx99 committed:
Update on "Make step() faster by passing in a tensor vs scalar 1"

This is the culminated result of https://github.com/pytorch/pytorch/pull/110954#issuecomment-1758520411.

We are making the code slightly more complicated to gain some perf in minimizing calls to `.copy_()` and `.to()`.

### Code
```
import torch
with torch.cuda.device(0):
    steps = [torch.zeros((), device="cpu", dtype=torch.float32) for i in range(1000)]

    with torch.profiler.profile(
        activities=[
            torch.profiler.ProfilerActivity.CPU,
            torch.profiler.ProfilerActivity.CUDA,
        ]
    ) as p:
        # New code:
        # step_device = steps[0].device
        # one = torch.tensor(1.0, device=step_device) if str(step_device) == "cpu" else 1
        # torch._foreach_add_(steps, one, 1.0)

        # Old code:
        torch._foreach_add_(steps, 1)

    print(p.key_averages().table(sort_by="cpu_time_total"))
```

### Profiles
**with old code**
```
-------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
                     Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg    # of Calls  
-------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
      aten::_foreach_add_        35.31%      52.089ms        99.99%     147.495ms     147.495ms             1  
               aten::add_        25.05%      36.949ms        64.68%      95.406ms      95.406us          1000  
                 aten::to         3.97%       5.852ms        39.63%      58.457ms      58.457us          1000  
           aten::_to_copy        10.11%      14.917ms        35.66%      52.605ms      52.605us          1000  
              aten::copy_        21.65%      31.939ms        21.65%      31.939ms      31.939us          1000  
      aten::empty_strided         3.90%       5.749ms         3.90%       5.749ms       5.749us          1000  
    cudaDeviceSynchronize         0.01%      18.000us         0.01%      18.000us      18.000us             1  
-------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
Self CPU time total: 147.513ms
```

**with new code**
```
-------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
                     Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg    # of Calls  
-------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
      aten::_foreach_add_        55.06%      49.963ms        99.86%      90.625ms      90.625ms             1  
               aten::add_        44.81%      40.662ms        44.81%      40.662ms      40.662us          1000  
            aten::detach_         0.01%       8.000us         0.05%      45.000us      45.000us             1  
                  detach_         0.04%      37.000us         0.04%      37.000us      37.000us             1  
              aten::empty         0.03%      30.000us         0.03%      30.000us      30.000us             1  
                 aten::to         0.03%      23.000us         0.03%      23.000us      23.000us             1  
    cudaDeviceSynchronize         0.02%      22.000us         0.02%      22.000us      22.000us             1  
         aten::lift_fresh         0.01%       6.000us         0.01%       6.000us       6.000us             1  
-------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
Self CPU time total: 90.751ms
```




[ghstack-poisoned]

janeyx99 commented on torch/optim/adagrad.py:
```diff
@@ -349,7 +349,12 @@ def _multi_tensor_adagrad(
+        # wrapped it now and passed it to the API as a tensor already. So we wrap it now.
+        step_device = device_state_steps[0].device
+        one = torch.tensor(1.0, device=step_device) if str(step_device) == "cpu" else 1
+        torch._foreach_add_(device_state_steps, one, 1.0)
```
For some reason, specifying an alpha (it could be 1.0 or just 1) is necessary for the perf win. I have no idea why the default of 1 is different and I'm a bit tired to look into this now. This is certainly interesting and deserves figuring out though.

awgu replied:
Should we check if alpha has any regression if "step" is on GPU? (assuming that "step" not being on CPU generally implies it is on GPU)

janeyx99 replied:
that’s a good idea, will do tomorrow

janeyx99 replied:
I'm gonna use an if else blatantly. I've also discovered the reason why the alpha specification is necessary, and it does not seem easily fixable: https://github.com/pytorch/pytorch/issues/111157

janeyx99 committed:
Update on "Make step() faster by passing in a tensor vs scalar 1"

This is the culminated result of https://github.com/pytorch/pytorch/pull/110954#issuecomment-1758520411.

We are making the code slightly more complicated to gain some perf in minimizing calls to `.copy_()` and `.to()`.

### Code
```
import torch
with torch.cuda.device(0):
    steps = [torch.zeros((), device="cpu", dtype=torch.float32) for i in range(1000)]

    with torch.profiler.profile(
        activities=[
            torch.profiler.ProfilerActivity.CPU,
            torch.profiler.ProfilerActivity.CUDA,
        ]
    ) as p:
        # New code:
        # step_device = steps[0].device
        # one = torch.tensor(1.0, device=step_device) if str(step_device) == "cpu" else 1
        # torch._foreach_add_(steps, one, 1.0)

        # Old code:
        torch._foreach_add_(steps, 1)

    print(p.key_averages().table(sort_by="cpu_time_total"))
```

### Profiles
**with old code**
```
-------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
                     Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg    # of Calls  
-------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
      aten::_foreach_add_        35.31%      52.089ms        99.99%     147.495ms     147.495ms             1  
               aten::add_        25.05%      36.949ms        64.68%      95.406ms      95.406us          1000  
                 aten::to         3.97%       5.852ms        39.63%      58.457ms      58.457us          1000  
           aten::_to_copy        10.11%      14.917ms        35.66%      52.605ms      52.605us          1000  
              aten::copy_        21.65%      31.939ms        21.65%      31.939ms      31.939us          1000  
      aten::empty_strided         3.90%       5.749ms         3.90%       5.749ms       5.749us          1000  
    cudaDeviceSynchronize         0.01%      18.000us         0.01%      18.000us      18.000us             1  
-------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
Self CPU time total: 147.513ms
```

**with new code**
```
-------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
                     Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg    # of Calls  
-------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
      aten::_foreach_add_        55.06%      49.963ms        99.86%      90.625ms      90.625ms             1  
               aten::add_        44.81%      40.662ms        44.81%      40.662ms      40.662us          1000  
            aten::detach_         0.01%       8.000us         0.05%      45.000us      45.000us             1  
                  detach_         0.04%      37.000us         0.04%      37.000us      37.000us             1  
              aten::empty         0.03%      30.000us         0.03%      30.000us      30.000us             1  
                 aten::to         0.03%      23.000us         0.03%      23.000us      23.000us             1  
    cudaDeviceSynchronize         0.02%      22.000us         0.02%      22.000us      22.000us             1  
         aten::lift_fresh         0.01%       6.000us         0.01%       6.000us       6.000us             1  
-------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
Self CPU time total: 90.751ms
```




[ghstack-poisoned]

albanD commented on torch/optim/adamax.py:
```diff
@@ -317,7 +317,12 @@ def _multi_tensor_adamax(
+        # wrapped it now and passed it to the API as a tensor already. So we wrap it now.
+        step_device = grouped_state_steps[0].device
+        one = torch.tensor(1.0, device=step_device) if str(step_device) == "cpu" else 1
+        torch._foreach_add_(grouped_state_steps, one, 1.0)
```
Why is this one different?

janeyx99 replied:
lol it shouldn't be

albanD commented on torch/optim/adagrad.py:
```diff
@@ -349,7 +349,13 @@ def _multi_tensor_adagrad(
+        # If steps are on CPU, foreach will fall back to the slow path, which is a for-loop calling t.add(1) over
+        # and over. 1 will then be wrapped into a Tensor over and over again, which is slower than if we just
+        # wrapped it now and passed it to the API as an already-made tensor. So we wrap it now.
+        if str(device_state_steps[0].device) == 'cpu':
```
```suggestion
        if device_state_steps[0].is_cpu:
```

janeyx99 replied:
:o crazy

albanD reviewed (approved):
nit on style but sgtm

janeyx99 committed:
Update on "Make step() faster by passing in a tensor vs scalar 1"

This is the culminated result of https://github.com/pytorch/pytorch/pull/110954#issuecomment-1758520411.

We are making the code slightly more complicated to gain some perf in minimizing calls to `.copy_()` and `.to()`.

### Code
```
import torch
with torch.cuda.device(0):
    steps = [torch.zeros((), device="cpu", dtype=torch.float32) for i in range(1000)]

    with torch.profiler.profile(
        activities=[
            torch.profiler.ProfilerActivity.CPU,
            torch.profiler.ProfilerActivity.CUDA,
        ]
    ) as p:
        # New code:
        # step_device = steps[0].device
        # one = torch.tensor(1.0, device=step_device) if str(step_device) == "cpu" else 1
        # torch._foreach_add_(steps, one, 1.0)

        # Old code:
        torch._foreach_add_(steps, 1)

    print(p.key_averages().table(sort_by="cpu_time_total"))
```

### Profiles
**with old code**
```
-------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
                     Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg    # of Calls  
-------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
      aten::_foreach_add_        35.31%      52.089ms        99.99%     147.495ms     147.495ms             1  
               aten::add_        25.05%      36.949ms        64.68%      95.406ms      95.406us          1000  
                 aten::to         3.97%       5.852ms        39.63%      58.457ms      58.457us          1000  
           aten::_to_copy        10.11%      14.917ms        35.66%      52.605ms      52.605us          1000  
              aten::copy_        21.65%      31.939ms        21.65%      31.939ms      31.939us          1000  
      aten::empty_strided         3.90%       5.749ms         3.90%       5.749ms       5.749us          1000  
    cudaDeviceSynchronize         0.01%      18.000us         0.01%      18.000us      18.000us             1  
-------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
Self CPU time total: 147.513ms
```

**with new code**
```
-------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
                     Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg    # of Calls  
-------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
      aten::_foreach_add_        55.06%      49.963ms        99.86%      90.625ms      90.625ms             1  
               aten::add_        44.81%      40.662ms        44.81%      40.662ms      40.662us          1000  
            aten::detach_         0.01%       8.000us         0.05%      45.000us      45.000us             1  
                  detach_         0.04%      37.000us         0.04%      37.000us      37.000us             1  
              aten::empty         0.03%      30.000us         0.03%      30.000us      30.000us             1  
                 aten::to         0.03%      23.000us         0.03%      23.000us      23.000us             1  
    cudaDeviceSynchronize         0.02%      22.000us         0.02%      22.000us      22.000us             1  
         aten::lift_fresh         0.01%       6.000us         0.01%       6.000us       6.000us             1  
-------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
Self CPU time total: 90.751ms
```




[ghstack-poisoned]

janeyx99 committed:
Update on "Make step() faster by passing in a tensor vs scalar 1"

This is the culminated result of https://github.com/pytorch/pytorch/pull/110954#issuecomment-1758520411.

We are making the code slightly more complicated to gain some perf in minimizing calls to `.copy_()` and `.to()`.

### Code
```
import torch
with torch.cuda.device(0):
    steps = [torch.zeros((), device="cpu", dtype=torch.float32) for i in range(1000)]

    with torch.profiler.profile(
        activities=[
            torch.profiler.ProfilerActivity.CPU,
            torch.profiler.ProfilerActivity.CUDA,
        ]
    ) as p:
        # New code:
        # step_device = steps[0].device
        # one = torch.tensor(1.0, device=step_device) if str(step_device) == "cpu" else 1
        # torch._foreach_add_(steps, one, 1.0)

        # Old code:
        torch._foreach_add_(steps, 1)

    print(p.key_averages().table(sort_by="cpu_time_total"))
```

### Profiles
**with old code**
```
-------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
                     Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg    # of Calls  
-------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
      aten::_foreach_add_        35.31%      52.089ms        99.99%     147.495ms     147.495ms             1  
               aten::add_        25.05%      36.949ms        64.68%      95.406ms      95.406us          1000  
                 aten::to         3.97%       5.852ms        39.63%      58.457ms      58.457us          1000  
           aten::_to_copy        10.11%      14.917ms        35.66%      52.605ms      52.605us          1000  
              aten::copy_        21.65%      31.939ms        21.65%      31.939ms      31.939us          1000  
      aten::empty_strided         3.90%       5.749ms         3.90%       5.749ms       5.749us          1000  
    cudaDeviceSynchronize         0.01%      18.000us         0.01%      18.000us      18.000us             1  
-------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
Self CPU time total: 147.513ms
```

**with new code**
```
-------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
                     Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg    # of Calls  
-------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
      aten::_foreach_add_        55.06%      49.963ms        99.86%      90.625ms      90.625ms             1  
               aten::add_        44.81%      40.662ms        44.81%      40.662ms      40.662us          1000  
            aten::detach_         0.01%       8.000us         0.05%      45.000us      45.000us             1  
                  detach_         0.04%      37.000us         0.04%      37.000us      37.000us             1  
              aten::empty         0.03%      30.000us         0.03%      30.000us      30.000us             1  
                 aten::to         0.03%      23.000us         0.03%      23.000us      23.000us             1  
    cudaDeviceSynchronize         0.02%      22.000us         0.02%      22.000us      22.000us             1  
         aten::lift_fresh         0.01%       6.000us         0.01%       6.000us       6.000us             1  
-------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
Self CPU time total: 90.751ms
```




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

janeyx99 commented on torch/_inductor/lowering.py:
```diff
@@ -225,7 +225,6 @@ def _register_foreach_lowering(aten_fn, decomp_fn):
 
     @functools.wraps(decomp_fn)
     def wrapped(*args, **kwargs):
-        assert len(args) <= 2
```
@mlazos I could change this to assert len(args) <= 3 to account for the alpha, but when I look into the lowered code, the alpha is nowhere to be found, so there might be a bigger problem here that should be addressed.

In this current PR, the alpha doesn't really matter, but I feel like either this assert should be turned to 3 OR something bigger may be happening. In the other make_pointwise code, alpha is handled by multiplying with the last set of inputs. I'm curious why we don't do that in the grouping code in make_pointwise_foreach.

mlazos replied:
This only supports a kwarg alpha it looks like, I don't think we can arbitrarily set the len(args) <= 3, there needs to be additional handling. I think the only alpha that is supported is through kwargs, I didn't even know there was a non-kwarg one tbh.

janeyx99 replied:
I re-added the assert, named the alpha, and tried running test_adagrad. It still fails on the assert with alpha=1 as the 3rd argument.
```
  File "/data/users/janeyx/pytorch/torch/_inductor/lowering.py", line 228, in wrapped
    assert len(args) <= 2
torch._dynamo.exc.BackendCompilerFailed: backend='inductor' raised:
LoweringException: AssertionError: 
  target: aten._foreach_add.Tensor
  args[0]: [TensorBox(StorageBox(
    InputBuffer(name='arg8_1', layout=FixedLayout('cpu', torch.float32, size=[], stride=[]))
  )), TensorBox(StorageBox(
    InputBuffer(name='arg9_1', layout=FixedLayout('cpu', torch.float32, size=[], stride=[]))
  )), TensorBox(StorageBox(
    InputBuffer(name='arg10_1', layout=FixedLayout('cpu', torch.float32, size=[], stride=[]))
  )), TensorBox(StorageBox(
    InputBuffer(name='arg11_1', layout=FixedLayout('cpu', torch.float32, size=[], stride=[]))
  ))]
  args[1]: TensorBox(StorageBox(
    Pointwise(
      'cpu',
      torch.float32,
      def inner_fn(index):
          tmp0 = ops.constant(1.0, torch.float32)
          return tmp0
      ,
      ranges=[],
      origin_node=full_default,
      origins={full_default}
    )
  ))
  args[2]: 1.0
```

mlazos replied:
This code works for me: https://gist.github.com/mlazos/77b1cb4cf69b147afebacbdef4469855

janeyx99 replied:
Resolved--I missed a * in the native_functions overload which made the alpha no longer kwarg-only. Thanks for pointing that out!

janeyx99 committed:
Update on "Make step() faster by passing in a tensor vs scalar 1"

This is the culminated result of https://github.com/pytorch/pytorch/pull/110954#issuecomment-1758520411.

We are making the code slightly more complicated to gain some perf in minimizing calls to `.copy_()` and `.to()`.

### Code
```
import torch
with torch.cuda.device(0):
    steps = [torch.zeros((), device="cpu", dtype=torch.float32) for i in range(1000)]

    with torch.profiler.profile(
        activities=[
            torch.profiler.ProfilerActivity.CPU,
            torch.profiler.ProfilerActivity.CUDA,
        ]
    ) as p:
        # New code:
        # step_device = steps[0].device
        # one = torch.tensor(1.0, device=step_device) if str(step_device) == "cpu" else 1
        # torch._foreach_add_(steps, one, 1.0)

        # Old code:
        torch._foreach_add_(steps, 1)

    print(p.key_averages().table(sort_by="cpu_time_total"))
```

### Profiles
**with old code**
```
-------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
                     Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg    # of Calls  
-------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
      aten::_foreach_add_        35.31%      52.089ms        99.99%     147.495ms     147.495ms             1  
               aten::add_        25.05%      36.949ms        64.68%      95.406ms      95.406us          1000  
                 aten::to         3.97%       5.852ms        39.63%      58.457ms      58.457us          1000  
           aten::_to_copy        10.11%      14.917ms        35.66%      52.605ms      52.605us          1000  
              aten::copy_        21.65%      31.939ms        21.65%      31.939ms      31.939us          1000  
      aten::empty_strided         3.90%       5.749ms         3.90%       5.749ms       5.749us          1000  
    cudaDeviceSynchronize         0.01%      18.000us         0.01%      18.000us      18.000us             1  
-------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
Self CPU time total: 147.513ms
```

**with new code**
```
-------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
                     Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg    # of Calls  
-------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
      aten::_foreach_add_        55.06%      49.963ms        99.86%      90.625ms      90.625ms             1  
               aten::add_        44.81%      40.662ms        44.81%      40.662ms      40.662us          1000  
            aten::detach_         0.01%       8.000us         0.05%      45.000us      45.000us             1  
                  detach_         0.04%      37.000us         0.04%      37.000us      37.000us             1  
              aten::empty         0.03%      30.000us         0.03%      30.000us      30.000us             1  
                 aten::to         0.03%      23.000us         0.03%      23.000us      23.000us             1  
    cudaDeviceSynchronize         0.02%      22.000us         0.02%      22.000us      22.000us             1  
         aten::lift_fresh         0.01%       6.000us         0.01%       6.000us       6.000us             1  
-------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
Self CPU time total: 90.751ms
```




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

janeyx99 committed:
Update on "Make step() faster by passing in a tensor vs scalar 1"

This is the culminated result of https://github.com/pytorch/pytorch/pull/110954#issuecomment-1758520411.

We are making the code slightly more complicated to gain some perf in minimizing calls to `.copy_()` and `.to()`.

### Code
```
import torch
with torch.cuda.device(0):
    steps = [torch.zeros((), device="cpu", dtype=torch.float32) for i in range(1000)]

    with torch.profiler.profile(
        activities=[
            torch.profiler.ProfilerActivity.CPU,
            torch.profiler.ProfilerActivity.CUDA,
        ]
    ) as p:
        # New code:
        # step_device = steps[0].device
        # one = torch.tensor(1.0, device=step_device) if str(step_device) == "cpu" else 1
        # torch._foreach_add_(steps, one, 1.0)

        # Old code:
        torch._foreach_add_(steps, 1)

    print(p.key_averages().table(sort_by="cpu_time_total"))
```

### Profiles
**with old code**
```
-------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
                     Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg    # of Calls  
-------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
      aten::_foreach_add_        35.31%      52.089ms        99.99%     147.495ms     147.495ms             1  
               aten::add_        25.05%      36.949ms        64.68%      95.406ms      95.406us          1000  
                 aten::to         3.97%       5.852ms        39.63%      58.457ms      58.457us          1000  
           aten::_to_copy        10.11%      14.917ms        35.66%      52.605ms      52.605us          1000  
              aten::copy_        21.65%      31.939ms        21.65%      31.939ms      31.939us          1000  
      aten::empty_strided         3.90%       5.749ms         3.90%       5.749ms       5.749us          1000  
    cudaDeviceSynchronize         0.01%      18.000us         0.01%      18.000us      18.000us             1  
-------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
Self CPU time total: 147.513ms
```

**with new code**
```
-------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
                     Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg    # of Calls  
-------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
      aten::_foreach_add_        55.06%      49.963ms        99.86%      90.625ms      90.625ms             1  
               aten::add_        44.81%      40.662ms        44.81%      40.662ms      40.662us          1000  
            aten::detach_         0.01%       8.000us         0.05%      45.000us      45.000us             1  
                  detach_         0.04%      37.000us         0.04%      37.000us      37.000us             1  
              aten::empty         0.03%      30.000us         0.03%      30.000us      30.000us             1  
                 aten::to         0.03%      23.000us         0.03%      23.000us      23.000us             1  
    cudaDeviceSynchronize         0.02%      22.000us         0.02%      22.000us      22.000us             1  
         aten::lift_fresh         0.01%       6.000us         0.01%       6.000us       6.000us             1  
-------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
Self CPU time total: 90.751ms
```




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

janeyx99 committed:
Update on "Make step() faster by passing in a tensor vs scalar 1"

This is the culminated result of https://github.com/pytorch/pytorch/pull/110954#issuecomment-1758520411.

We are making the code slightly more complicated to gain some perf in minimizing calls to `.copy_()` and `.to()`.

### Code
```
import torch
with torch.cuda.device(0):
    steps = [torch.zeros((), device="cpu", dtype=torch.float32) for i in range(1000)]

    with torch.profiler.profile(
        activities=[
            torch.profiler.ProfilerActivity.CPU,
            torch.profiler.ProfilerActivity.CUDA,
        ]
    ) as p:
        # New code:
        # step_device = steps[0].device
        # one = torch.tensor(1.0, device=step_device) if str(step_device) == "cpu" else 1
        # torch._foreach_add_(steps, one, 1.0)

        # Old code:
        torch._foreach_add_(steps, 1)

    print(p.key_averages().table(sort_by="cpu_time_total"))
```

### Profiles
**with old code**
```
-------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
                     Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg    # of Calls  
-------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
      aten::_foreach_add_        35.31%      52.089ms        99.99%     147.495ms     147.495ms             1  
               aten::add_        25.05%      36.949ms        64.68%      95.406ms      95.406us          1000  
                 aten::to         3.97%       5.852ms        39.63%      58.457ms      58.457us          1000  
           aten::_to_copy        10.11%      14.917ms        35.66%      52.605ms      52.605us          1000  
              aten::copy_        21.65%      31.939ms        21.65%      31.939ms      31.939us          1000  
      aten::empty_strided         3.90%       5.749ms         3.90%       5.749ms       5.749us          1000  
    cudaDeviceSynchronize         0.01%      18.000us         0.01%      18.000us      18.000us             1  
-------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
Self CPU time total: 147.513ms
```

**with new code**
```
-------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
                     Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg    # of Calls  
-------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
      aten::_foreach_add_        55.06%      49.963ms        99.86%      90.625ms      90.625ms             1  
               aten::add_        44.81%      40.662ms        44.81%      40.662ms      40.662us          1000  
            aten::detach_         0.01%       8.000us         0.05%      45.000us      45.000us             1  
                  detach_         0.04%      37.000us         0.04%      37.000us      37.000us             1  
              aten::empty         0.03%      30.000us         0.03%      30.000us      30.000us             1  
                 aten::to         0.03%      23.000us         0.03%      23.000us      23.000us             1  
    cudaDeviceSynchronize         0.02%      22.000us         0.02%      22.000us      22.000us             1  
         aten::lift_fresh         0.01%       6.000us         0.01%       6.000us       6.000us             1  
-------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
Self CPU time total: 90.751ms
```




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

mlazos commented on torch/_inductor/lowering.py:
```diff
@@ -4733,6 +4733,7 @@ def register_pointwise_numeric_ldf64(op):
 
 register_foreach_pointwise(aten._foreach_add.List, add, allow_alpha=True)
 register_foreach_pointwise(aten._foreach_add.Scalar, add, allow_alpha=True)
+register_foreach_pointwise(aten._foreach_add.Tensor, add, allow_alpha=True)
```
If we are adding this registration can you add a test as well in test_foreach.py

janeyx99 replied:
Do you mean for eager? If so, the test is already added in the earlier PR of this stack (where the overload is added)

mlazos replied:
No I mean for inductor

janeyx99 replied:
Done in #111600

janeyx99 committed:
Update on "Make step() faster by passing in a tensor vs scalar 1"

This is the culminated result of https://github.com/pytorch/pytorch/pull/110954#issuecomment-1758520411.

We are making the code slightly more complicated to gain some perf in minimizing calls to `.copy_()` and `.to()`.

### Code
```
import torch
with torch.cuda.device(0):
    steps = [torch.zeros((), device="cpu", dtype=torch.float32) for i in range(1000)]

    with torch.profiler.profile(
        activities=[
            torch.profiler.ProfilerActivity.CPU,
            torch.profiler.ProfilerActivity.CUDA,
        ]
    ) as p:
        # New code:
        # step_device = steps[0].device
        # one = torch.tensor(1.0, device=step_device) if str(step_device) == "cpu" else 1
        # torch._foreach_add_(steps, one, 1.0)

        # Old code:
        torch._foreach_add_(steps, 1)

    print(p.key_averages().table(sort_by="cpu_time_total"))
```

### Profiles
**with old code**
```
-------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
                     Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg    # of Calls  
-------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
      aten::_foreach_add_        35.31%      52.089ms        99.99%     147.495ms     147.495ms             1  
               aten::add_        25.05%      36.949ms        64.68%      95.406ms      95.406us          1000  
                 aten::to         3.97%       5.852ms        39.63%      58.457ms      58.457us          1000  
           aten::_to_copy        10.11%      14.917ms        35.66%      52.605ms      52.605us          1000  
              aten::copy_        21.65%      31.939ms        21.65%      31.939ms      31.939us          1000  
      aten::empty_strided         3.90%       5.749ms         3.90%       5.749ms       5.749us          1000  
    cudaDeviceSynchronize         0.01%      18.000us         0.01%      18.000us      18.000us             1  
-------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
Self CPU time total: 147.513ms
```

**with new code**
```
-------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
                     Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg    # of Calls  
-------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
      aten::_foreach_add_        55.06%      49.963ms        99.86%      90.625ms      90.625ms             1  
               aten::add_        44.81%      40.662ms        44.81%      40.662ms      40.662us          1000  
            aten::detach_         0.01%       8.000us         0.05%      45.000us      45.000us             1  
                  detach_         0.04%      37.000us         0.04%      37.000us      37.000us             1  
              aten::empty         0.03%      30.000us         0.03%      30.000us      30.000us             1  
                 aten::to         0.03%      23.000us         0.03%      23.000us      23.000us             1  
    cudaDeviceSynchronize         0.02%      22.000us         0.02%      22.000us      22.000us             1  
         aten::lift_fresh         0.01%       6.000us         0.01%       6.000us       6.000us             1  
-------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
Self CPU time total: 90.751ms
```




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

janeyx99 commented:
@pytorchbot merge
